ironic/resource_ironic_node_v1.go

```diff
@@ -585,3 +621,105 @@ func changePowerState(client *gophercloud.ServiceClient, d *schema.ResourceData,
 
 	return nil
 }
+
+// setRAIDConfig calls ironic's API to send request to change a Node's RAID config.
+func setRAIDConfig(client *gophercloud.ServiceClient, d *schema.ResourceData) (err error) {
+	var logicalDisks []nodes.LogicalDisk
+	var raid *metal3v1alpha1.RAIDConfig
+
+	raidConfig := d.Get("raid_config").(string)
+	if raidConfig == "" {
+		return nil
+	}
+
+	err = json.Unmarshal([]byte(raidConfig), &raid)
+	if err != nil {
+		return
+	}
+
+	err = checkRAIDConfigure(d.Get("raid_interface").(string), raid)
+	if err != nil {
+		return
+	}
+
+	// Build target for RAID configuration steps
+	logicalDisks, err = ironic.BuildTargetRAIDCfg(raid)
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+
+	// Set root volume
+	if len(d.Get("root_device").(map[string]interface{})) == 0 {
+		logicalDisks[0].IsRootVolume = new(bool)
+		*logicalDisks[0].IsRootVolume = true
+	} else {
+		log.Printf("rootDeviceHints is used, the first volume of raid will not be set to root")
+	}
+
+	// Set target for RAID configuration steps
+	return nodes.SetRAIDConfig(
+		client,
+		d.Id(),
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+}
+
+// buildManualCleaningSteps builds the clean steps for RAID and BIOS configuration
+func buildManualCleaningSteps(d *schema.ResourceData) (cleanSteps []nodes.CleanStep, err error) {
+	var targetRaid *metal3v1alpha1.RAIDConfig
+	var settings []map[string]string
+
+	raidInterface := d.Get("raid_interface").(string)
+
+	raidConfig := d.Get("raid_config").(string)
+	if raidConfig != "" {
+		if err = json.Unmarshal([]byte(raidConfig), &targetRaid); err != nil {
+			return nil, err
+		}
+
+		// Build raid clean steps
+		raidCleanSteps, err := ironic.BuildRAIDCleanSteps(raidInterface, targetRaid, nil)
+		if err != nil {
+			return nil, err
+		}
+		cleanSteps = append(cleanSteps, raidCleanSteps...)
+	}
+
+	biosSetings := d.Get("bios_settings").(string)
+	if biosSetings != "" {
+		if err = json.Unmarshal([]byte(biosSetings), &settings); err != nil {
+			return nil, err
+		}
+
+		cleanSteps = append(
+			cleanSteps,
+			nodes.CleanStep{
+				Interface: "bios",
+				Step:      "apply_configuration",
+				Args: map[string]interface{}{
+					"settings": settings,
+				},
+			},
+		)
+	}
+
+	return
+}
+
+func checkRAIDConfigure(raidInterface string, raid *metal3v1alpha1.RAIDConfig) error {
```

Comment:
The logic here is wrong, see https://github.com/metal3-io/baremetal-operator/pull/1063. I wonder if we can avoid duplication of BMO code here. We can make that function public.

Reply:
> The logic here is wrong, see https://github.com/metal3-io/baremetal-operator/pull/1063.

Thanks, we have [commented](https://github.com/metal3-io/baremetal-operator/pull/1063#issuecomment-1034407549) under PR1063.

> I wonder if we can avoid duplication of BMO code here. We can make that function public.

That would be nice.

Reply:
I've created a PR that only exposes the function without fixing it: https://github.com/metal3-io/baremetal-operator/pull/1094. This way you can proceed for now.

Reply:
That's great. Thanks a lot!